File: go/alpha.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/tls"
	"encoding/base64"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
	"time"
)

func normalize(value string) string {
	// Percent-encode a string, excluding unreserved characters.
	return url.QueryEscape(value)
}

func formatTimestamp(t time.Time) string {
	// Format timestamp as 'yyyy-MM-dd'T'HH:mm:ss.SSS'Z''
	return t.UTC().Format("2006-01-02T15:04:05.000Z")
}

func hmacSHA256(key []byte, message string) []byte {
	// Compute HMAC-SHA256 and return the raw digest.
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(message))
	return mac.Sum(nil)
}

func buildCanonicalQueryString(queryParameters map[string]string) string {
	if len(queryParameters) == 0 {
		return ""
	}
	// Sort the query parameters
	var keys []string
	for k := range queryParameters {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var queryParts []string
	for _, k := range keys {
		v := queryParameters[k]
		encodedKey := normalize(k)
		encodedValue := normalize(v)
		queryParts = append(queryParts, fmt.Sprintf("%s=%s", encodedKey, encodedValue))
	}
	return strings.Join(queryParts, "&")
}

func buildCanonicalRequest(httpMethod, uri string, queryParameters map[string]string, signedHeaders []string, canonicalHeaders map[string]string, payload string) string {
	var cr strings.Builder
	cr.WriteString(httpMethod)
	cr.WriteString("\n")
	cr.WriteString(uri)
	cr.WriteString("\n")

	// Canonical query string
	canonicalQueryString := buildCanonicalQueryString(queryParameters)
	cr.WriteString(canonicalQueryString)
	cr.WriteString("\n")

	// Signed headers
	cr.WriteString(strings.Join(signedHeaders, ";"))
	cr.WriteString("\n")

	// Canonical headers
	for _, headerName := range signedHeaders {
		headerValue := canonicalHeaders[headerName]
		cr.WriteString(fmt.Sprintf("%s:%s\n", headerName, headerValue))
	}
	cr.WriteString("\n")

	// Normalized payload
	normalizedPayload := normalize(payload)
	cr.WriteString(normalizedPayload)

	return cr.String()
}

func main() {
	// Inputs
	accessKey := "globalaktest"
	secretKey := "1qaz2wsx3edc4rfv5tgb6yhn7ujm8ik9ol0p"
	host := "10.22.26.181:28080"
	uri := "/rest/cmsapp/v1/ping"
	httpMethod := "POST"
	timestamp := time.Now().UTC()
	formattedTimestamp := formatTimestamp(timestamp)
	payload := `{"say": "Hello world!"}`

	// Example query parameters (if any)
	queryParameters := map[string]string{
		// "param1": "value1",
	}

	// Compute Content-Length
	contentLength := fmt.Sprintf("%d", len(payload))

	// Construct headers
	headers := map[string]string{
		"host":           host,
		"content-length": contentLength,
		"content-type":   "application/json;charset=UTF-8",
	}

	// Lowercase and sort header names
	var signedHeaders []string
	for k := range headers {
		signedHeaders = append(signedHeaders, strings.ToLower(k))
	}
	sort.Strings(signedHeaders)

	// Canonical headers with normalized values
	canonicalHeaders := make(map[string]string)
	for _, k := range signedHeaders {
		v := headers[k]
		normalizedValue := normalize(strings.TrimSpace(v))
		canonicalHeaders[k] = normalizedValue
	}

	// Build CanonicalRequest
	canonicalRequest := buildCanonicalRequest(
		httpMethod, uri, queryParameters, signedHeaders, canonicalHeaders, payload,
	)

	// Build authStringPrefix
	authVersion := "auth-v2"
	signedHeadersStr := strings.Join(signedHeaders, ";")
	authStringPrefix := fmt.Sprintf("%s/%s/%s/%s", authVersion, accessKey, formattedTimestamp, signedHeadersStr)

	// Compute SigningKey (raw bytes)
	signingKey := hmacSHA256([]byte(secretKey), authStringPrefix)

	// Compute Signature (raw bytes)
	signatureBytes := hmacSHA256(signingKey, canonicalRequest)

	// Base64-encode the signature
	signature := base64.StdEncoding.EncodeToString(signatureBytes)

	// Build Authorization header
	authorization := fmt.Sprintf("%s/%s", authStringPrefix, signature)

	// Construct final headers for the request
	requestHeaders := map[string]string{
		"Authorization":  authorization,
		"Host":           headers["host"],
		"Content-Length": headers["content-length"],
		"Content-Type":   headers["content-type"],
	}

	// Create HTTP client with proxy
	proxyURL, err := url.Parse("http://proxy.example.com:8080") // Replace with your proxy URL
	if err != nil {
		fmt.Println("Invalid proxy URL:", err)
		os.Exit(1)
	}

	transport := &http.Transport{
		Proxy: http.ProxyURL(proxyURL),
		// Skip SSL verification (not recommended for production)
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		// Dialer with timeout
		DialContext: (&net.Dialer{
			Timeout: 30 * time.Second,
		}).DialContext,
	}

	client := &http.Client{
		Transport: transport,
		Timeout:   60 * time.Second,
	}

	// Build the request
	page := fmt.Sprintf("https://%s%s", host, uri)
	req, err := http.NewRequest(httpMethod, page, strings.NewReader(payload))
	if err != nil {
		fmt.Println("Error creating request:", err)
		return
	}

	// Add query parameters to the URL if any
	if len(queryParameters) > 0 {
		q := req.URL.Query()
		for k, v := range queryParameters {
			q.Add(k, v)
		}
		req.URL.RawQuery = q.Encode()
	}

	// Set headers
	for k, v := range requestHeaders {
		req.Header.Set(k, v)
	}

	// Send the request
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Request error:", err)
		return
	}
	defer func() {
		if err := resp.Body.Close(); err != nil {
			fmt.Printf("Error closing response body: %v\n", err)
		}
	}()

	// Read the response
	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response:", err)
		return
	}

	// Print the response
	fmt.Println("\nResponse Status Code:")
	fmt.Println(resp.StatusCode)
	fmt.Println("\nResponse Headers:")
	for k, v := range resp.Header {
		fmt.Printf("%s: %s\n", k, strings.Join(v, ", "))
	}
	fmt.Println("\nResponse Body:")
	fmt.Println(string(bodyBytes))
}

File: python/alpha.py
import hashlib
import hmac
import urllib.parse
from datetime import datetime, timezone
import base64
import requests  # Ensure you have requests installed: pip install requests


def normalize(value):
    """
    Percent-encode a string, excluding unreserved characters.
    """
    return urllib.parse.quote(value, safe='-._~')


def format_timestamp(timestamp):
    """
    Format the timestamp as 'yyyy-MM-dd'T'HH:mm:ss.SSS'Z''.
    """
    return timestamp.strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z'


def hmac_sha256(key, msg):
    """
    Compute HMAC-SHA256 and return the raw bytes.
    """
    if isinstance(key, str):
        key = key.encode('utf-8')
    return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()


def build_canonical_query_string(query_parameters):
    """
    Build the canonical query string by normalizing and sorting the query parameters.
    """
    if not query_parameters:
        return ''
    # Sort the query parameters alphabetically
    sorted_params = sorted((normalize(k), normalize(v)) for k, v in query_parameters.items())
    # Concatenate the sorted parameters
    canonical_query_string = '&'.join(f"{k}={v}" for k, v in sorted_params)
    return canonical_query_string


def build_canonical_request(http_method, uri, query_parameters, signed_headers, canonical_headers, payload):
    """
    Construct the CanonicalRequest string.
    """
    cr = [http_method, '\n', uri, '\n']

    # Canonical query string
    canonical_query_string = build_canonical_query_string(query_parameters)
    cr.append(canonical_query_string)
    cr.append('\n')

    # Signed headers
    cr.append(';'.join(signed_headers))
    cr.append('\n')

    # Canonical headers
    for header_name in signed_headers:
        header_value = canonical_headers[header_name]
        cr.append(f'{header_name}:{header_value}\n')
    cr.append('\n')

    # Normalized payload
    normalized_payload = normalize(payload)
    cr.append(normalized_payload)

    return ''.join(cr)


def main():
    # Inputs
    access_key = "globalaktest"
    secret_key = "1qaz2wsx3edc4rfv5tgb6yhn7ujm8ik9ol0p"
    host = "10.22.26.181:28080"
    uri = "/rest/cmsapp/v1/ping"
    http_method = "POST"
    timestamp = datetime.now(timezone.utc)
    formatted_timestamp = format_timestamp(timestamp)
    payload = '{"say": "Hello world!"}'

    # Example query parameters (if any)
    query_parameters = {
        # 'param1': 'value1',
        # 'param2': 'value2',
    }

    # Compute Content-Length
    content_length = str(len(payload.encode('utf-8')))

    # Construct headers
    headers = {
        'host': host,
        'content-length': content_length,
        'content-type': 'application/json;charset=UTF-8'
    }

    # Lowercase and sort header names
    signed_headers = sorted([k.lower() for k in headers.keys()])

    # Canonical headers with normalized values
    canonical_headers = {}
    for k in signed_headers:
        v = headers[k]
        normalized_value = normalize(v.strip())
        canonical_headers[k] = normalized_value

    # Build CanonicalRequest
    canonical_request = build_canonical_request(
        http_method, uri, query_parameters, signed_headers, canonical_headers, payload
    )

    # Build authStringPrefix
    auth_version = "auth-v2"
    signed_headers_str = ';'.join(signed_headers)
    auth_string_prefix = f"{auth_version}/{access_key}/{formatted_timestamp}/{signed_headers_str}"

    # Compute SigningKey (raw bytes)
    signing_key = hmac_sha256(secret_key, auth_string_prefix)

    # Compute Signature (raw bytes)
    signature_digest = hmac_sha256(signing_key, canonical_request)

    # Base64-encode the signature
    signature = base64.b64encode(signature_digest).decode('utf-8')

    # Build Authorization header
    authorization = f"{auth_string_prefix}/{signature}"

    # Construct final headers for the request
    request_headers = {
        'Authorization': authorization,
        'Host': headers['host'],
        'Content-Length': headers['content-length'],
        'Content-Type': headers['content-type'],
    }

    # Define the proxy (replace with your proxy details)
    proxies = {
        'http': 'http://proxy.example.com:8080',  # Replace with your proxy URL
        'https': 'http://proxy.example.com:8080',  # Replace with your proxy URL
    }

    # Prepare the full URL
    url = f"https://{host}{uri}"

    # Send the HTTP POST request through the proxy
    response = requests.post(
        url,
        headers=request_headers,
        data=payload,
        proxies=proxies,
        verify=False,  # Set to True in production
        timeout=30
    )

    # Print the response
    print("\nResponse Status Code:")
    print(response.status_code)
    print("\nResponse Headers:")
    for k, v in response.headers.items():
        print(f"{k}: {v}")
    print("\nResponse Body:")
    print(response.text)


if __name__ == "__main__":
    main()
